client/verta/verta/runtime.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+
+import json
+from typing import Any, Dict
+
+
+class LogHandler:
+    """
+    Container class for collecting and dumping log entries.
+    """
+    def __init__(self):
+        self.attributes: Dict[str, Any] = dict()
+
+    def __repr__(self):
+        return str(self.attributes)
+
+    def add(self, **kwargs):
+        # Overwrites existing key-values
+        self.attributes.update(**kwargs)
+
+    def wipe_clean(self):
+        self.attributes = dict()
+
+# Use a single handler object that is wiped clean after each log entry
+# to minimize memory usage
+LOG_HANDLER = LogHandler()
+
+
+def log(**kwargs) -> None:
+    """
+    Log any number of key-value pairs to the current prediction context.
+    Values are deleted upon completion of the active model's predict() method.
+
+    Parameters
+    ----------
+    kwargs : Any
+        Arguments and values to be added as key-value pairs to the current log record.
+        Values must be serializable to JSON.
+
+    Returns
+    -------
+    None
+    """
+    LOG_HANDLER.add(**kwargs)
+
+
+class ContextLogger():
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    Any logged data is deleted upon exiting the context.
+
+    Attributes
+    ----------
+    as_json : str
+        A JSON string representation of all the current saved logging context.
+    """
+    def __init__(self):
+        self.context = LOG_HANDLER
+        if bool(self.context.attributes): # if not empty dict
+            self.context.wipe_clean()
+
+    def log(self, **kwargs) -> None:
+        self.context.add(**kwargs)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        # Handler attributes get deleted when context is exited.
+        self.context.wipe_clean()
+
+    @property
+    def as_json(self) -> str:
```

Comment:
Isn't this supposed to be

```suggestion
    def logs(self) -> dict:
```

as per the design doc we agreed on?

https://github.com/VertaAI/docs/blob/main/design-docs/2023-01-model-prediction-logging.md#logging-key-values

I don't see mention of any reason to have deviated from that.

Reply:
Also, as a personal opinion, I don't think this should be a `property` either way. I believe that if a user accesses an object attribute:

```python
ctx.as_json
```

they should almost _never_ expect to encounter an exception (other than a `object has no attribute` `AttributeError`). If an exception can occur due to extra internal processing (or moreso due to data validation), it should be a method call:

```python
ctx.as_json()
```

See https://github.com/psf/requests/issues/1028#issuecomment-11474800

Reply:
Fair point.  I agree with your logic here. ~Fixed~

Reply:
Fixed naming convention to `logs()`

Reply:
It doesn't look like we've fully addressed my suggestion of

```python
def logs(self) -> dict:
```

The [design doc](https://github.com/VertaAI/docs/blob/main/design-docs/2023-01-model-prediction-logging.md#logging-key-values) outlines that `ctx.logs()` should return a `dict`, not a `str`.

If you are intent on changing this to a `str`, I **highly** recommend raising it #feat-model-data-log to get input from others. Otherwise, this breaks the API that we had consensus on for everyone consuming this downstream.

Reply:
Got it.  I thought we had discussed in standup whether it should return a dict or a JSON string and I understood JSON to be the answer.  Easy thing to change.

Reply:
Refactored to return `Dict[str, Any]`

Reply:
Ah, a misunderstanding, then. Sorry about the confusion.